### DISCUSSION TITLE
Allow for custom server for frontend apps

### DISCUSSION THREAD

**Discussion by DennisGaida** [ente-io/ente#504] [Enhancements]
Since the server is now [open-sourced](https://ente.io/blog/open-sourcing-our-server/) and can be hosted using Docker, I would love to keep using the existing frontend apps (Android, iOS & Windows), but use my own backend / URL.

I can do that when building everything myself, but it is quite a hassle - taking iOS as an example with needing an Apple developer account and a developer enabled iOS device etc.

I want to be able to use the default consumer apps, but just be able to switch to a custom server. I'm especially interested in **ente Auth** having this feature. The official Bitwarden app has exactly that feature - you can self-host the Bitwarden server and just use its URL in the official app instead of the commercially hosted Bitwarden version.

Feature request in short: I want to be able to use my custom server URL / backend in the existing frontend apps for iOS, Android & Windows.

**Comment by ioslife** [comment]
This would be huge. I'm super excited about getting the chance to selfhost this and compare it to Immich!

**Reply by davidSGE** [reply]
Same here!

**Comment by arthurlt** [comment]
I’ve landed on just using Synology Photos, but I’m uncomfortable with the theoretical access I have to everyone’s photos as the admin. Plus, the backup experience on iOS leaves _a lot_ to be desired.

I’m super interested in pivoting my family to using this when this feature is implemented. 

Side note: If I didn’t already have replicated local storage I would be interested in the paid hosted solution. Are there any ways to recurrently donate? You’re providing a unique feature set.

**Reply by natebc** [reply]
Github Sponsorship is active for this project, You could donate there.

**Comment by tarunkumar519** [comment]
This feature would be awesome, super interested

**Comment by bdonvr** [comment]
I'm super, super interested in this. I'm selfhosting everything else and would really like to be able to selfhost my 2FA app. As soon as we can have an easy to deploy server and web app (docker?), and the ability to use a custom server URI without building the app from source I'll be thrilled.

**Reply by DennisGaida** [reply]
We do have an *easy to deploy server and web app* in docker. That is the reason I created this issue because now we're only missing the frontend apps (aside the web app).

**Reply by monarc83** [reply]
is the docker already available for public somewhere? That would be great. Thanks.,

**Reply by qdlmcfresh** [reply]
Have a look at: https://github.com/ente-io/ente/blob/main/server/README.md

**Reply by monarc83** [reply]
Thanks for the quick reply, but to be honest, I think I need more of a step by step guide with some more explanations for better understanding. If there is something like that already available somewhere, it would be great if someone can point me to it or maybe someone is willing to write this guide.

**Comment by aravindhsampath** [comment]
Since this feature would amount to letting self hosters running the whole thing for free - as in beer, I would even be happy if the folks behind the project would make a copy of the app titled specifically as “ente - selfhosted” or something like that, and charge a reasonable fee for it. 

This way, self hosters can pay for the convenience of the app, contributing financially to the project, while still using their own server and photo storage. 

Count me as your customer #1 if you get there. 

I played around with the ente app, in view of replacing iCloud Photos. But, the lack of ability to use the app to my own backend stopped me from proceeding any further. If I need to pay a cloud vendor anyway, I might as well pay Apple and enjoy the integration convenience.

**Reply by immanuelfodor** [reply]
I think an opposing opinion should also be expressed and heard, not just the "shut up and take my money" one, so I'm adding a counter argument :grinning: 

I'd be happy if it could stay free as in beer for self-hosting folks. The huge majority of users probably wouldn't be able to self-host it anyways, and those who can, probably can be considered power users. And power users are usually enthusiastic about the project, can do beta test, provide feedback, report bugs, open feature requests, etc. so the project can benefit from them even if  not charging for the self-hosted deployment. 

However, maybe it would need tiers to not let companies self-host it for free, so a differentiation of self-hosters. Like my home usage could go free with up to ~3-5 users, but any more users than that could go with a self-hosted license. But I wouldn't lock the feature availability behind a paid license. 

I'm looking for something that can replace my personal Google Photos which I'm using for free, and this project looks like the best alternative so far. I'd go with a private self-hosted install, but if it'd require payment, I'd immediately unsubscribe from watching the repo which I started because of the news of going open-sourced and the possibility of self-hosting the whole stack.

**Reply by tarunkumar519** [reply]
as @immanuelfodor said opening the apps to self-hosting will give the software lot of traction with many new users who may also convert to paid subs, new PRs, new ideas and good community kicking in. I have seen this benefiting a lot of good open source projects for a long time.

**Comment by primeinc** [comment]
Considering the importance of cost and data management, perhaps exploring a **hybrid storage model** would be beneficial. This model could allow for **archiving less critical photos on personal servers,** while keeping frequently accessed ones on Ente's cloud. 

It acknowledges the practicality of self-hosting for archival data, ensuring seamless integration with Ente's cloud for essential access, even when the self-hosted servers are temporarily offline. This approach promotes data autonomy and cost savings, aligning with our shared goals of efficiency and control over our digital footprint.

**Comment by vishnukvmd** [comment]
**Update**: We will first allow Auth to connect to a self-hosted instance.

This necessary changes are done in https://github.com/ente-io/ente/pull/726, and will be available in the next release.

Photos will follow.

**Reply by DennisGaida** [reply]
Whoop whoop!!!!! I‘ll be sure to join in on #726 - maybe I will be able to help in some way.

**Reply by FarisZR** [reply]
Any updates on when photos will support this?

**Reply by mnvr** [reply]
We're waiting for things to stablize a bit, just seeing how it is going with this change in auth and taking into account any feedback we're getting before we make the same change in photos. So no ETA, but we'll add this, and we'll do it sooner than later!

**Reply by jo-chemla** [reply]
Hi there, jumping on the bandwagon @mnvr and @vishnukvmd: do you plan to release the built ente-server docker container image - eg to github container registry ghcr.io or docker-hub? This way self-hosters would spawn the ente-stack even more easily within a standard docker-compose environment. Thanks and congrats for for the great work!

**Reply by mnvr** [reply]
Yes, this is planned! See https://github.com/ente-io/ente/discussions/673#discussioncomment-8828341, tl;dr; is that we'll soon provide prebuilt ente-server images too.

**Reply by jo-chemla** [reply]
Thanks, that's great news! Good server [readme] (https://github.com/ente-io/ente/blob/main/server%2FREADME.md) by the way! 
- so once the server image build is published, the stack required to run it would also include
- a postgres as database 
- regarding the storage, where would things reside? 

I've seen in your [reliability post](https://ente.io/reliability/) that your team uses backblaze and wasabi as hot and scaleway as cold storage. Would self-hosters have to rely on one of these or will they also be able to self-host their data in an u opinionated way? Thanks again!

**Reply by mnvr** [reply]
> Good server [readme]

Thank you!

> so once the server image build is published, the stack required to run it would also include a postgres as database, regarding the storage, where would things reside?

I feel there is some confusion. If you wish for a full stack solution (ente's server + postgres + storage), that is already possible. The server directory has a docker compose file, and you can just clone the repo, `cd server`, and do a `docker compose up` to get all three of them.

The reason for publishing an additional standalone Docker image for ente's server is twofold:

* People can use this compose file without needing to clone the repository
* But more importantly, some people just want Ente's server, and they wish to setup their own Postgres and S3-compatible storage. For such folks, having a Docker image would remove any need for them to take anything from the repository.

> Would self-hosters have to rely on one of these or will they also be able to self-host their data in an u opinionated way?

You can use whatever you wish, it is totally unopinionated. As long as the service you're using provides an S3-compatible API, it can be used with Ente's server. Our starter docker compose file uses MinIO, and on our production deployment we use B2/Wasabi/SCW as you mentioned. You can use these, or anything in between. See [help.ente.io/self-hosting](https://help.ente.io/self-hosting/) for some guides etc around this.

Hope this helps!

**Reply by tarunkumar519** [reply]
Thank you for clearing some confusion. Self-hosting using MinIO will be good. But is it possible to store files in directly without any object storage? If this deviates from how ente is built & if not possible i understand.

**Reply by mnvr** [reply]
>  is it possible to store files in directly without any object storage?

Ah. Well, no, that's not possible yet.

**Reply by mnvr** [reply]
> is it possible to store files in directly without any object storage?

I was pondering over, trying to think of the use case you might have in mind. Perhaps you want a "raw" backup of your files, without necessarily structuring them how they get stored in the S3 compatible object storage. Well, in that case, it's already possible, many customers do it. You can use Ente's [CLI](https://github.com/ente-io/ente/tree/main/cli#ente-cli) to setup cron jobs to run `ente export`, or use Ente's [desktop app](https://ente.io/download/desktop) to setup automatic and continuous exports on your machine. 

That way, you don't need to care how your self hosted instance is storing the data in MinIO etc. 

1. Your app connects to your self hosted Ente, say using the starter compose file.
2. Self hosted Ente stores everything as a blackbox.
3. You run the Ente CLI/desktop app to run continuous exports to some other drive.

In such a setup, your self hosted Ente will be a disposable bit, since your primary backup will be the export that results from the export step.

Not sure if this is what you had in mind, but I thought I'll explain this approach anyway, might help someone. As I mentioned, many customers already do this (step 2 doesn't care if it is self hosted Ente or our paid production service).

**Reply by codethief** [reply]
@mnvr Thanks, that indeed clarifies things a bit!

> In such a setup, your self hosted Ente will be a disposable bit, since your primary backup will be the export that results from the export step.

Is there any documentation on what exactly `ente export` does and, in particular, on the format of the exports that it creates? What data will be included? (Background: I already have an elaborate folder structure for my photos and would rather not lose that.) Also, is there an inverse "import" command, or is the desktop app the only way currently to move stuff to Ente?

**Reply by mnvr** [reply]
1. It'll export all your photos/videos, in the exact same folder structure that they are being shown in the Ente Photos app. Since we don't use compression, you'll get back the data in its original fidelity. We'll try to also include as much of the metadata you added within Ente app in JSON files in the export - currently this is not 100%, and it'll probably not make sense in some cases (e.g. what photo is shared with whom is not something that can be readily exported), but we do try to vend out as much as we can, and we keep adding more to this.

2. You're right, there isn't an inverse import command (though now I'm wondering why not! 😄 more seriously though, the import is a one time thing, the export is the more critical bit to be automated which is why we added that first).

**Reply by natebc** [reply]
> I was pondering over, trying to think of the use case you might have in mind. 

In my mind, the use-case without object storage is just ... storing the files onto local storage. Think running the container and just writing the files out to a filesystem instead of shipping them off to s3 of some variety.

If one is already self-hosting an object storage server then this isn't a huge undertaking to just configure Ente to use that. However, not everyone is running their own object storage server (or may not have a use for one other than this). In that instance being able to just pass a volume of some type into the container would be useful.

**Reply by mkdirmushroom** [reply]
@mnvr  Thank you for the detailed explanation on the ente export functionality. One aspect I wanted to clarify - since the export creates a full backup each time, as more files accumulate in Ente over time, won't the export file size keep growing larger with each backup? If so, at some point, the export files may become quite large.

Do you have any recommendations or best practices around managing this growth in export file size? For example, are there plans to introduce an incremental export option in the future? Or any other strategies you'd suggest to handle the increasing export data as an Ente library expands? I'd appreciate your thoughts on effectively managing the export process long-term.

**Reply by mnvr** [reply]
The exports are already incremental! Both CLI and desktop app will just resume exports from where they left off, and/or export newer files, whenever you restart the export process. So you don't have to worry about the export size growing unnecessarily.

**Reply by mkdirmushroom** [reply]
Great, thank you very much!

**Reply by mkdirmushroom** [reply]
@natebc This may go against their original intentions, considering security and privacy, the use of object storage is understandable.

**Reply by aravindhsampath** [reply]
I don’t want miss an update that brings the ability for ente photos iOS app to connect to a self hosted backend. Is there any place I can watch for that action?

**Reply by mnvr** [reply]
You can watch this discussion (you'll already be watching it since you commented here). We'll make sure we announce here when that change is done.

EDIT: GitHub will notify you I mean when there is a new comment on this discussion.

**Reply by jo-chemla** [reply]
Thanks @mnvr for the clarifications! Indeed what I had in mind as a selfhoster was indeed to host both server (+db, storage) and webapp, I'll watch the other threads (like [this one](https://github.com/ente-io/ente/issues/983#issuecomment-2055849502)) regarding the progress on publishing an ente photos/auth webapp.

**Comment by wayneyaoo** [comment]
Hey thank you very much for the great product!! Not sure if I should start a separate discussion on this but, a dumb question...  does Ente Auth server share the same backend with Ente Photo? I mean when I look at the dir structure I only see `museum`. Did I miss something? Can I self-host only the 2FA auth server?

**Reply by mnvr** [reply]
Yes, they share the same backend. And yes, you can self-host only the 2FA auth server. The starter Docker compose will work fine for either Photos or Auth (or both!) ☺️

(I think a separate discussion would've been better, many people are watching this particular discussion and they might be getting spammed by unrelated posts 😬)

**Reply by wayneyaoo** [reply]
Yes I understand and I've got the answer. Thank you very much! Let me mute and not spam people :)...

**Comment by DevBefell** [comment]
Not sure if already a thing but since implementing a custom server for the frontends it also would be great for us developers to have some debugging features (which can be enabled via the same way of the changing the server), whether logs etc.

**Reply by mnvr** [reply]
Off the top of my head - It isn't really something we're too keen on doing. We don't want ente to become a kitchen sink. But if you feel it's useful, please create a new discussion for this! We'll see if it gets some traction, and implement the needful (I'm just trying to keep this discussion scoped, so that watchers don't get unnecessary notifications).

**Comment by vishnukvmd** [comment]
**Update**: Photos app now supports connecting to a self-hosted instance.

The changes were made in https://github.com/ente-io/ente/pull/1302 and the release is available here: https://github.com/ente-io/ente/releases/tag/photos-v0.8.75

Will keep this issue open until the release is live on PlayStore, AppStore & F-Droid.

**Reply by vishnukvmd** [reply]
Just fyi - our GitHub releases support auto-updates

**Reply by the-hotmann** [reply]
Seems like the update is out ^^

**Reply by wjbeckett** [reply]
Will this extend to the desktop app too?

**Reply by mnvr** [reply]
It isn't planned, the web and desktop apps are much easier to build compared to building say the iOS app which requires Xcode, so with mobile apps this was needed more. 

That said, we'll add it to this option to the desktop app too once we get a bit of time, will keep this in mind.

**Comment by wjbeckett** [comment]
> It isn't planned, the web and desktop apps are much easier to build compared to building say the iOS app which requires Xcode, so with mobile apps this was needed more.
> 
> That said, we'll add it to this option to the desktop app too once we get a bit of time, will keep this in mind.

Fair enough. Do self-hosters have the ability to bulk upload though? I thought that was only achievable via the Desktop app?

**Reply by mnvr** [reply]
Self hosters can always build a desktop app and bulk upload, there is no restriction. This feature request was only for making that even easier. 

As I said, I understand the need.

**Reply by vishnukvmd** [reply]
@wjbeckett FYI: Our web app supports bulk upload as well. The important feature the desktop app offers for uploads is parsing of zipped Google Takeouts.

So if you're importing from your hard disk, the web app should do just fine!

**Reply by wjbeckett** [reply]
Thanks for the info.
Just waiting for the web app to be containerised :-)

**Reply by jpeeler** [reply]
Is the web app planning to be containerized? I'm not able to get it to run, but will patiently wait if it's on the roadmap somewhere.

**Reply by mnvr** [reply]
There are two different things - 

1. Allowing connecting to a custom self hosted instance from the desktop app. This is on the cards.

2. Containerizing the web app. This is a different feature request, e.g. see here - https://github.com/ente-io/ente/issues/983. This too is on the roadmap - we'd like people to be able to directly run a docker cluster that has the web app and the server itself, a sort of a "demo" instance (https://github.com/ente-io/ente/discussions/982), but I'm not sure when we'll be able to carve out time to get around to doing it. In the linked issue there are some comments and docker files that already do it, but nothing point-and-click yet.

> Re running the web app itself, these might be useful (just in case you haven't seen these yet): https://help.ente.io/self-hosting/guides/web-app

**Comment by TesterEU** [comment]
Hello,
I'm sorry that I didn't manage to read the entire discussion, but I've a question (I'm a non-technical person):
Is it possible to run the ente application on a shared hosting provider (I've purchased an account)? I don't have a VPS, and my hosting has php, mysql, composer (not to be confused with git). I'd like to test Your server application because it seems great ;) I currently have Piwigo installed.

**Reply by mnvr** [reply]
This thread is not the correct place to ask general self hosting questions, you end up notifying all the people who're watching it for updates on the original issue. If people keep doing this we'll have to lock this thread.

About your question itself - If you're a non-technical person, I generally would not recommend trying to self host, it is not in a state where it is seamless enough (we're hoping to improve that, cf see this https://github.com/ente-io/ente/issues/983#issuecomment-2054808064). Immich, for example, might be a better option. 

About the shared hosting provider - From your description it doesn't seem like they give the ability to run Docker images, in that case unfortunately you won't be able to self host on that provider. It sounds like they setup they provide is for PHP apps.

**Comment by mnvr** [comment]
Update: The desktop app (and also self hosted web app) now supports connecting to a self-hosted instance (the same way it works in the mobile apps).

The changes are available in the latest version, v1.7.2. For more details, see https://help.ente.io/self-hosting/guides/custom-server/#desktop-and-web.

All clients now support connecting to custom servers, as OP had requested, so I'll now close this feature request. Thank you all for your patience.
